private/bufpkg/bufcheck/buflint/internal/buflintvalidate/cel.go

```diff
@@ -118,19 +99,120 @@ func validateCELCompilesField(
 	if err != nil {
 		return err
 	}
-	for i, celConstraint := range fieldConstraints.GetCel() {
-		celLocation := field.OptionExtensionLocation(
-			validate.E_Field,
-			celFieldTagInFieldConstraints,
-			int32(i),
-		)
-		checkCel(celEnv, celConstraint, func(format string, args ...interface{}) {
-			add(field, celLocation, nil, format, args...)
-		})
-	}
+	checkCEL(
+		celEnv,
+		fieldConstraints.GetCel(),
+		fmt.Sprintf("field %q", adder.fieldName()),
+		fmt.Sprintf("Field %q", adder.fieldName()),
+		adder.getFieldRuleName(celFieldNumberInFieldConstraints),
+		func(index int, format string, args ...interface{}) {
+			adder.addForPathf(
+				[]int32{celFieldNumberInFieldConstraints, int32(index)},
+				format,
+				args...,
+			)
+		},
+	)
 	return nil
 }
 
+func checkCEL(
+	celEnv *cel.Env,
+	celConstraints []*validate.Constraint,
+	parentName string,
+	parentNameCapitalized string,
+	celName string,
+	add func(int, string, ...interface{}),
+) {
+	idToConstraintIndices := make(map[string][]int, len(celConstraints))
+	for i, celConstraint := range celConstraints {
+		if celID := celConstraint.GetId(); celID != "" {
+			for _, char := range celID {
+				if 'a' <= char && char <= 'z' {
+					continue
+				} else if 'A' <= char && char <= 'Z' {
+					continue
+				} else if '0' <= char && char <= '9' {
+					continue
+				} else if char == '_' || char == '-' || char == '.' {
+					continue
+				}
+				add(
+					i,
+					"%s has an invalid %s.id, an id must contain only alphanumerics and '.', '_' and '-'.",
```

Comment:
`%s has invalid characters for %s.id. IDs must contain only characters a-z, A-Z, 0-9, '.', '_', '-'.`